### PR TITLE
feat(connector): [HELCIM] Move connector to hyperswitch_connectors

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/5289

## How did you test it?

Following flows need to be tested for connector Helcim
1. Authorize:
Request:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_EvozwXY09n7tTs0SbWXKKGuEBIAtODBGwPTeu4iDlCx8izBWqyOF5QCNjPccy5qV' \
--data '{
    "amount": 1742,
    "currency": "USD",
    "confirm": true,
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "payment_method": "card",
    "payment_method_type": "credit",
    "setup_future_usage": "on_session",
    "payment_method_data": {
        "card": {
            "card_number": "5413330089099130",
            "card_exp_month": "01",
            "card_exp_year": "27",
            "card_cvc": "123",
            "card_holder_name": "John Doe"
        }
    },
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "billing": {
        "address": {
            "line1": "casd",
            "zip": "94122",
            "first_name": "John",
            "last_name": "Doe"
        }
    },
    "customer_id": "chethan"
}'
```
Response:
```
{
    "payment_id": "pay_TCUn15IkjNK5RVJvhkVQ",
    "merchant_id": "merchant_1721640890",
    "status": "succeeded",
    "amount": 1742,
    "net_amount": 1742,
    "amount_capturable": 0,
    "amount_received": 1742,
    "connector": "helcim",
    "client_secret": "pay_TCUn15IkjNK5RVJvhkVQ_secret_vw61UKv9fMjFtLLIqvxP",
    "created": "2024-07-24T09:52:40.945Z",
    "currency": "USD",
    "customer_id": "chethan",
    "customer": {
        "id": "chethan",
        "name": null,
        "email": null,
        "phone": null,
        "phone_country_code": null
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "on_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": null,
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "9130",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "541333",
            "card_extended_bin": null,
            "card_exp_month": "01",
            "card_exp_year": "27",
            "card_holder_name": "John Doe",
            "payment_checks": null,
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": null,
            "country": null,
            "line1": "casd",
            "line2": null,
            "line3": null,
            "zip": "94122",
            "state": null,
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": null,
        "email": null
    },
    "order_details": null,
    "email": null,
    "name": null,
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "chethan",
        "created_at": 1721814760,
        "expires": 1721818360,
        "secret": "epk_fff23eaae15744eda3e21f25da60d9c5"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "27216935",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_TCUn15IkjNK5RVJvhkVQ_1",
    "payment_link": null,
    "profile_id": "pro_LFHne2pOYzWQinLzkwCx",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_x6fFNn94BtH7q72XabAM",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-07-24T10:07:40.944Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-07-24T09:52:43.689Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null
}
```

2.  Capture:
Request:
```
curl --location 'http://localhost:8080/payments/pay_ocZh7RY06eyLv1cRi4O5/capture' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_EvozwXY09n7tTs0SbWXKKGuEBIAtODBGwPTeu4iDlCx8izBWqyOF5QCNjPccy5qV' \
--data '{
}'
```

Response:
```
{
    "payment_id": "pay_ocZh7RY06eyLv1cRi4O5",
    "merchant_id": "merchant_1721640890",
    "status": "succeeded",
    "amount": 1742,
    "net_amount": 1742,
    "amount_capturable": 0,
    "amount_received": 1742,
    "connector": "helcim",
    "client_secret": "pay_ocZh7RY06eyLv1cRi4O5_secret_w59OCpVfwMQtbfFEN7Nd",
    "created": "2024-07-24T09:54:19.891Z",
    "currency": "USD",
    "customer_id": "chethan",
    "customer": {
        "id": "chethan",
        "name": null,
        "email": null,
        "phone": null,
        "phone_country_code": null
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "on_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "9130",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "541333",
            "card_extended_bin": null,
            "card_exp_month": "01",
            "card_exp_year": "27",
            "card_holder_name": "John Doe",
            "payment_checks": null,
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": null,
            "country": null,
            "line1": "casd",
            "line2": null,
            "line3": null,
            "zip": "94122",
            "state": null,
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": null,
        "email": null
    },
    "order_details": null,
    "email": null,
    "name": null,
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "27216940",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "",
    "payment_link": null,
    "profile_id": "pro_LFHne2pOYzWQinLzkwCx",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_x6fFNn94BtH7q72XabAM",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-07-24T10:09:19.891Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-07-24T09:54:34.462Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/hyperswitch_connectors/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+error-stack = "0.4.1"
+serde = { version = "1.0.197", features = ["derive"] }
+serde_json = "1.0.115"
+
+# First party crates
+api_models = { version = "0.1.0", path = "../api_models", features = ["errors"] }
+cards = { version = "0.1.0", path = "../cards" }
+common_enums = { version = "0.1.0", path = "../common_enums" }
+common_utils = { version = "0.1.0", path = "../common_utils", features = ["signals", "async_ext", "logs", "metrics"] }
+diesel_models = { version = "0.1.0", path = "../diesel_models", features = ["kv_store"] }
```
Comment (jarnura) [inline crates/hyperswitch_connectors/Cargo.toml:19]: what are the dependency connected to diesel_models?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/hyperswitch_connectors/Cargo.toml (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/Cargo.toml]: The dependency was not being used directly and is now removed.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/hyperswitch_connectors/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    RouterData<Authorize, PaymentsAuthorizeData, PaymentsResponseData>;
+pub type PaymentsSyncRouterData = RouterData<PSync, PaymentsSyncData, PaymentsResponseData>;
+pub type PaymentsCaptureRouterData = RouterData<Capture, PaymentsCaptureData, PaymentsResponseData>;
+pub type PaymentsCancelRouterData = RouterData<Void, PaymentsCancelData, PaymentsResponseData>;
+pub type SetupMandateRouterData =
+    RouterData<SetupMandate, SetupMandateRequestData, PaymentsResponseData>;
+pub type RefundsRouterData<F> = RouterData<F, RefundsData, RefundsResponseData>;
+pub type RefundSyncRouterData = RouterData<RSync, RefundsData, RefundsResponseData>;
+pub type RefundsResponseRouterData<F, R> =
+    ResponseRouterData<F, R, RefundsData, RefundsResponseData>;
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/types.rs:21]: Moved these types except `RefundsResponseRouterData` as discussed.

File: crates/hyperswitch_connectors/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+license.workspace = true
+
+[features]
+frm = ["hyperswitch_domain_models/frm"]
+payouts = ["hyperswitch_domain_models/payouts", "api_models/payouts"]
+
+[dependencies]
+async-trait = "0.1.79"
+error-stack = "0.4.1"
+nanoid = "0.4.0"
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/Cargo.toml:15]: Done

File: crates/hyperswitch_connectors/src/constants.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+#[rustfmt::skip]
+pub(crate) const ALPHABETS: [char; 62] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+    'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+    'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+];
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/constants.rs:45]: Done

File: crates/hyperswitch_connectors/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,1084 @@
...
+}
+
+#[inline]
+pub fn generate_id(length: usize, prefix: &str) -> String {
+    format!(
+        "{}_{}",
+        prefix,
+        nanoid!(length, &crate::constants::ALPHABETS)
+    )
+}
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/utils.rs:88]: Done

File: crates/hyperswitch_connectors/src/constants.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
+/// Header Constants
+pub mod headers {
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/constants.rs:2]: Done

File: crates/hyperswitch_connectors/Cargo.toml
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/Cargo.toml:1]: Done

File: crates/hyperswitch_connectors/src/types.rs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+pub struct ResponseRouterData<Flow, R, Request, Response> {
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/types.rs:23]: Issue created for the same: https://github.com/juspay/hyperswitch-cloud/issues/6101

File: crates/hyperswitch_connectors/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,1084 @@
...
+use common_utils::{
+    errors::ReportSwitchExt,
+    ext_traits::{OptionExt, ValueExt},
+    id_type,
+    pii::{self, Email, IpAddress},
+};
+use error_stack::ResultExt;
+use hyperswitch_interfaces::{api, errors};
+use nanoid::nanoid;
+type Error = error_stack::Report<errors::ConnectorError>;
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/utils.rs:11]: Done

File: crates/hyperswitch_connectors/src/utils.rs
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/src/utils.rs:1]: Utility functions and types are made `pub(crate)`.

Regarding traits if we make them `pub(crate)` then we start getting dead code warning for the unused functions. We can either suppress this warning or we can make the traits as `pub(crate)` after moving all the connectors.

File: crates/hyperswitch_connectors/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+[dependencies]
+async-trait = "0.1.79"
+error-stack = "0.4.1"
+serde = { version = "1.0.197", features = ["derive"] }
+serde_json = "1.0.115"
+strum = { version = "0.26", features = ["derive"] }
+
+# First party crates
+api_models = { version = "0.1.0", path = "../api_models", features = ["errors"] , default-features = false}
+cards = { version = "0.1.0", path = "../cards" }
```
Comment (jarnura) [inline crates/hyperswitch_connectors/Cargo.toml:21]: Remove this dependency if unused

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/hyperswitch_connectors/Cargo.toml (hunk 9 above)
Comment (deepanshu-iiitu) [reply crates/hyperswitch_connectors/Cargo.toml:21]: This is being used in `helcim::transformers`

**Review by jarnura — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Pull request merged**